### DISCUSSION TITLE
Beatmaps unplayable due to increased input delay when rapidly pressing keys

### DISCUSSION THREAD

**Discussion by ricky8k** [ppy/osu#16756] [Q&A] [answered]
Reproduce:
1. Use keyboard as input
2. Hit any key rapidly

Doing this causes input latency to spike considerably, freezing other input such as moving the mouse and resulting in missed inputs. It appears the issue occurs even outside of beatmaps, as demonstrated by the video attached, where it occurs on the main menu (see the Input graph).

Issue does not exist under osu!stable, so it might be an osu!lazer issue.

Pressing Z/X (set as main input in settings) rapidly:

https://user-images.githubusercontent.com/92121005/152272090-cd2ce9db-0a0e-4702-936f-be5861ef8c37.mp4

Pressing any set of keys (as well as attempting to move the mouse):

https://user-images.githubusercontent.com/92121005/152272860-cc1d931c-91a0-4fc8-9a7a-e9bde93063a5.mp4

Running on Windows 10 21H1 (19043.1466).
Game is set at 4x refresh rate and multithreaded performance is enabled, though lowering/increasing the frame cap as well as switching to singlethreaded has not resolved this issue.
Problem has persisted since build 2021.1225.0, currently on 2022.128.0 at the time of writing (latest)

**Comment by peppy** [comment]
Something on your PC is intercepting your input and adding delay (white on input thread means another app is causing the latency). You will need to figure out what it is.

Please let us know when you find out which app.

**Reply by ricky8k** [reply]
I'll try to look into it tomorrow if I have time. Will keep you posted if I'm able to find the culprit, though I was confused as to why the issue isn't existent under osu!stable if it is indeed related to a 3rd-party software on my computer.

Two programs I know off the top of my head that control my keyboard behavior: DREVO Power Console (RGB and key mapping for my DREVO BladeMaster TE) and Keyboard Chattering Fix (eliminates extra inputs within certain ms timeframe), I assume it could be the latter if you're right.

**Reply by peppy** [reply]
There's any number of reasons it might not affect stable.

- stable is architected very differently, and may avoid the issue for one reason or another.
- the app which is causing the issue could have a blacklist where it doesn't apply to some apps.
- something could be hooking apps and choose which app to hook based on the gl context version or any other reason
- your stable and lazer installs may be running at different display modes

I could list a heap more reasons, but at the end of the day it's not too important - finding the cause should be the primary goal. I do have an idea of how we can potentially improve this going forward, but as it is quite a rare occurrence (affects few users) I haven't had the time to investigate further yet.

**Comment by ricky8k** [comment, marked as answer]
Apologies for the delay.

After taking a look into specific applications running on my system, I've pinpointed the cause of it, in that my Keyboard Chattering Fix utility was indeed the problem whenever left in the background, after a certain amount of time.
I can confirm that this issue is reproducible with this app after a few attempts: the issue is not present whenever the program is freshly launched and is running in the background, but after a couple of minutes the issue starts again, in which the input delay will increase. Closing the program will remove this delay.

You can see the video here, where I demonstrate the input delay being present, which is then resolved after closing the aforementioned program:

https://user-images.githubusercontent.com/92121005/152619233-03f91e34-b1de-4dc3-a606-f8f71092278b.mp4

(osu!lazer was put into Borderless for this demonstration, effect is still present in Fullscreen)

While this could be a superficial problem that will not affect most (if even any), if you wish to dig into it, I've linked a download to the program used here (if you trust that what I've attached is not a virus):
- [kcfix0.0.1.zip](https://github.com/ppy/osu/files/8006933/kcfix0.0.1.zip) (fresh copy, this does not have my config)

![image](https://user-images.githubusercontent.com/92121005/152628837-857fb327-468d-46f5-aafc-acf769d24081.png)


…though this probably means I need to look for an alternative to my key chattering problem.

**Reply by bdach** [reply]
these "anti-chattering" apps are probably using system-level hooks to intercept and delay key presses before they reach any actual applications. so no surprise they will cause "input lag". they just can't work much differently.

**Reply by ricky8k** [reply]
The anti-chattering software was definitely something that I suspected of causing the problem, but the issue could've been correlated to how dated the Keyboard Chattering Fix application is, and how it grabs key inputs being different from other, newer applications out there.

I'm currently trying a different key-chattering removal software that is much more recent, you can find the GitHub repo if you wish to look into it here:
- https://github.com/mcmonkeyprojects/KeyboardChatterBlocker
![image](https://user-images.githubusercontent.com/92121005/152647791-c1af15da-de23-43b8-841d-d1c8c3dec984.png)

…and so far, osu!lazer has not experienced any problems in terms of delay even after relaunching the program several times, as well as running it in the background for a long period. There is no increase in input delay or white lines which appear on the performance graph. It's placed under the same ms threshold (90ms) compared to the Keyboard Chattering Fix program. 

So, it might just be that the software was using an outdated method to grab input (it's from 2014, so I wouldn't be surprised).
